EKF/ekf_helper.cpp

```diff
@@ -1819,3 +1816,137 @@ void Ekf::stopFlowFusion()
 	memset(_flow_innov_var,0.0f,sizeof(_flow_innov_var));
 	memset(&_optflow_test_ratio,0.0f,sizeof(_optflow_test_ratio));
 }
+
+// Reset main nav filter yaw to value from EKF-GSF and reset velocity and position to GPS
+bool Ekf::resetYawToEKFGSF()
+{
+	// don't allow reet using the EKF-GSF estimate until the filter has started fusing velocity
+	// data and the yaw estimate has converged
+	float new_yaw, new_yaw_variance;
+	if (yawEstimator.getYawData(&new_yaw, &new_yaw_variance) && new_yaw_variance < sq(_params.EKFGSF_yaw_err_max)) {
+		// save a copy of the quaternion state for later use in calculating the amount of reset change
+		Quatf quat_before_reset = _state.quat_nominal;
+
+		// update transformation matrix from body to world frame using the current estimate
+		_R_to_earth = Dcmf(_state.quat_nominal);
+
+		// update the rotation matrix using the new yaw value
+		// determine if a 321 or 312 Euler sequence is best
+		if (fabsf(_R_to_earth(2, 0)) < fabsf(_R_to_earth(2, 1))) {
+			// use a 321 sequence
+			Eulerf euler321(_state.quat_nominal);
+			euler321(2) = new_yaw;
+			_R_to_earth = euler321;
+
+		} else {
+			// Calculate the 312 Tait-Bryan rotation sequence that rotates from earth to body frame
+			// We use a 312 sequence as an alternate when there is more pitch tilt than roll tilt
+			// to avoid gimbal lock
+			Vector3f rot312;
+			rot312(0) = new_yaw;
+			rot312(1) = asinf(_R_to_earth(2, 1));
+			rot312(2) = atan2f(-_R_to_earth(2, 0), _R_to_earth(2, 2));
+
+			float c2 = cosf(rot312(2));
+			float s2 = sinf(rot312(2));
+			float s1 = sinf(rot312(1));
+			float c1 = cosf(rot312(1));
+			float s0 = sinf(rot312(0));
+			float c0 = cosf(rot312(0));
+
+			_R_to_earth(0, 0) = c0 * c2 - s0 * s1 * s2;
+			_R_to_earth(1, 1) = c0 * c1;
+			_R_to_earth(2, 2) = c2 * c1;
+			_R_to_earth(0, 1) = -c1 * s0;
+			_R_to_earth(0, 2) = s2 * c0 + c2 * s1 * s0;
+			_R_to_earth(1, 0) = c2 * s0 + s2 * s1 * c0;
+			_R_to_earth(1, 2) = s0 * s2 - s1 * c0 * c2;
+			_R_to_earth(2, 0) = -s2 * c1;
+			_R_to_earth(2, 1) = s1;
```

Comment:
This looks the same as 
- https://github.com/PX4/ecl/blob/3fa5f501ae7edc4a793f0973e156706afcc8b419/EKF/mag_fusion.cpp#L650-L665
- https://github.com/PX4/ecl/blob/3fa5f501ae7edc4a793f0973e156706afcc8b419/EKF/gps_yaw_fusion.cpp#L344-L360
- https://github.com/PX4/ecl/blob/3fa5f501ae7edc4a793f0973e156706afcc8b419/EKF/ekf_helper.cpp#L600-L627

Can we package that 312 to DCM into a helper function to avoid a fourth duplicate?

Reply:
Done